oneflow/core/kernel/image_decoder_random_crop_resize_kernel.cpp

```diff
@@ -0,0 +1,479 @@
+  nvjpegImage_t image;
+  image.channel[0] = dst;
+  image.pitch[0] = roi_width * kNumChannels;
+  OF_NVJPEG_CHECK(nvjpegDecodeParamsSetOutputFormat(jpeg_decode_params_, NVJPEG_OUTPUT_RGBI));
+  OF_NVJPEG_CHECK(
+      nvjpegDecodeParamsSetROI(jpeg_decode_params_, roi_x, roi_y, roi_width, roi_height));
+  OF_NVJPEG_CHECK(nvjpegStateAttachPinnedBuffer(jpeg_state_, jpeg_pinned_buffer_));
+  OF_NVJPEG_CHECK(nvjpegStateAttachDeviceBuffer(jpeg_state_, jpeg_device_buffer_));
+  OF_NVJPEG_CHECK(nvjpegDecodeJpegHost(jpeg_handle_, jpeg_decoder_, jpeg_state_,
+                                       jpeg_decode_params_, jpeg_stream_));
+  OF_NVJPEG_CHECK(nvjpegDecodeJpegTransferToDevice(jpeg_handle_, jpeg_decoder_, jpeg_state_,
+                                                   jpeg_stream_, cuda_stream_));
+  OF_NVJPEG_CHECK(
+      nvjpegDecodeJpegDevice(jpeg_handle_, jpeg_decoder_, jpeg_state_, &image, cuda_stream_));
+  *dst_width = roi_width;
+  *dst_height = roi_height;
+}
+
+void GpuDecodeHandle::Decode(const unsigned char* data, size_t length, unsigned char* dst,
+                             size_t dst_max_length, int* dst_width, int* dst_height) {
+  DecodeRandomCrop(data, length, nullptr, dst, dst_max_length, dst_width, dst_height);
+}
+
+void GpuDecodeHandle::Resize(const unsigned char* src, int src_width, int src_height,
+                             unsigned char* dst, int dst_width, int dst_height) {
+  const NppiSize src_size{
+      .width = src_width, .height = src_height,
+  };
+  const NppiRect src_rect{
+      .x = 0, .y = 0, .width = src_width, .height = src_height,
+  };
+  const NppiSize dst_size{
+      .width = dst_width, .height = dst_height,
+  };
+  const NppiRect dst_rect{
+      .x = 0, .y = 0, .width = dst_width, .height = dst_height,
+  };
+  NppStatus status =
+      nppiResize_8u_C3R_Ctx(src, src_width * kNumChannels, src_size, src_rect, dst, dst_width * 3,
+                            dst_size, dst_rect, NPPI_INTER_LINEAR, npp_stream_ctx_);
+  CHECK_GE(status, NPP_SUCCESS);
+}
+
+void GpuDecodeHandle::DecodeRandomCropResize(const unsigned char* data, size_t length,
+                                             RandomCropGenerator* crop_generator,
+                                             unsigned char* workspace, size_t workspace_size,
+                                             unsigned char* dst, int target_width,
+                                             int target_height) {
+  int width;
+  int height;
+  DecodeRandomCrop(data, length, crop_generator, workspace, workspace_size, &width, &height);
+  Resize(workspace, width, height, dst, target_width, target_height);
+}
+
+void GpuDecodeHandle::WarmupOnce(int warmup_size, unsigned char* workspace, size_t workspace_size) {
+  if (warmup_done_) { return; }
+  warmup_size = std::min(static_cast<int>(std::sqrt(workspace_size / kNumChannels)), warmup_size);
+  cv::Mat image = cv::Mat::zeros(cv::Size(warmup_size, warmup_size), CV_8UC3);
```

Comment:
这里的逻辑是创建一个(warmup_size， warmup_size, 3)大小的全零图片，然后尝试decode？

Reply:
这个warmup size是用户配置的？我见是conf里传进来的。用户怎么知道要传多大？

Reply:
还是说这个size其实可以随便设置，然后在训练过程中，只要遇到了更大的图片，就会触发重新malloc的逻辑？

Reply:
看了一下DALI的文档，DALI是设置的16M显存，一般就够用了

```
device_memory_padding (int, optional, default = 16777216) –

`mixed` backend only Padding for nvJPEG’s device memory allocations in bytes. This parameter helps to avoid reallocation in nvJPEG whenever a bigger image is encountered and the internal buffer needs to be reallocated to decode it.
```

链接： https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.ops.ImageDecoder

Reply:
> 看了一下DALI的文档，DALI是设置的16M显存，一般就够用了
> 
> ```
> device_memory_padding (int, optional, default = 16777216) –
> 
> `mixed` backend only Padding for nvJPEG’s device memory allocations 
> in bytes. This parameter helps to avoid reallocation in nvJPEG whenever 
> a bigger image is encountered and the internal buffer needs to be 
> reallocated to decode it.
> ```
> 
> 链接： https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.ops.ImageDecoder

这个是指分配显存的时候按照 device_memory_padding 来申请，避免频繁的重分配

处理imagenet数据集的时候，单张图片最大需要120M左右的显存

> 还是说这个size其实可以随便设置，然后在训练过程中，只要遇到了更大的图片，就会触发重新malloc的逻辑？

是的